### ISSUE TITLE
2.0.36 breaks Codeception tests

### ISSUE THREAD

**Issue by DmLapin** [yiisoft/yii2#18245]
### What steps will reproduce the problem?
After install 2.0.36 update all our codeception tests became broken. Reason: https://github.com/yiisoft/yii2/pull/18127.
If I remove
```php
elseif (is_array($dependency)) {
    $dependencies[$index] = $this->resolveDependencies($dependency, $reflection);
}
```
from \yii\di\Container::resolveDependencies, tests become OK.

First of all, it **IS** breaking change a-priory because of `Container::resolveDependencies` behaviour change.

Problem starts here: `vendor/codeception/codeception/src/Codeception/Lib/Connector/Yii2.php:75`. Connector use `Yii::createObject` (and therefore dependency container) for Application creation. We have application config like this:
```php
return [
...
'container' => [
    'definitions' => [
        common\interfaces\NewsletterEmailGeneratorInterface::class => [
            ['class' => common\components\newsletter\NewsletterEmailGenerator::class],
            [
                di\Instance::of(web\View::class),
                di\Instance::of('internalFilesystem'),
]
...
```
Recursive `resolveDependencies` call try get 'internalFilesystem' component, but **it's not configured in DI Container on this stage!** We got:
`[yii\base\InvalidConfigException] Failed to instantiate component or class "internalFilesystem".`

Example of stack trace:
```php
 Test  unit/commands/news/news_domain/NewsBorrowCommandTest.php:testInvalidData

  [yii\base\InvalidConfigException] Failed to instantiate component or class "internalFilesystem".

.../vendor/yiisoft/yii2/di/Container.php:449
.../vendor/yiisoft/yii2/di/Container.php:447
.../vendor/yiisoft/yii2/di/Container.php:374
.../vendor/yiisoft/yii2/di/Container.php:159
.../vendor/yiisoft/yii2/di/Container.php:484
.../vendor/yiisoft/yii2/di/Container.php:491
.../vendor/yiisoft/yii2/di/Container.php:491
.../vendor/yiisoft/yii2/di/Container.php:491
.../vendor/yiisoft/yii2/di/Container.php:491
.../vendor/yiisoft/yii2/di/Container.php:395
.../vendor/yiisoft/yii2/di/Container.php:159
.../vendor/yiisoft/yii2/BaseYii.php:365
.../vendor/codeception/codeception/src/Codeception/Lib/Connector/Yii2.php:75
.../vendor/codeception/codeception/src/Codeception/Lib/Connector/Yii2.php:53
.../vendor/codeception/codeception/src/Codeception/Module/Yii2.php:185
.../vendor/codeception/codeception/src/Codeception/Subscriber/Module.php:56
```

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.36
| PHP version      | 7.4

**Comment by samdark** [comment]
> it's not configured in DI Container on this stage!

Where/how is it configured?

**Comment by DmLapin** [comment]
> Where/how is it configured?

Not sure, that understand question. When we create application in typical way (`$application = new yii\web\Application($config);`), container options from $config['container'] go to `\yii\base\Application::setContainer()`. But codeception creates application in another way:
```
/** @var \yii\web\Application $app */
$this->app = Yii::createObject($config);
```
See this part of stacktrace:
```
.../vendor/yiisoft/yii2/di/Container.php:395
.../vendor/yiisoft/yii2/di/Container.php:159
.../vendor/yiisoft/yii2/BaseYii.php:365
.../vendor/codeception/codeception/src/Codeception/Lib/Connector/Yii2.php:75
```
There is no configured application instance (and therefore no configured container) on line Container.php:395, but recursive dependency resolve starts.

**Comment by samdark** [comment]
I see. @SamMousa is it better to be adjusted in Codeception module or in the framework (not sure how yet)?

**Comment by SamMousa** [comment]
First of all:  #18127 seems to be bad regardless of this specific issue:
- As @DmLapin noted it changes behavior, so it is a BC break
- It introduces (possible) infinite recursion without checks

Then there is the Yii2 issue of attempting to have everything in the configuration array. Things like the logger, errorhandlers and possibly the container get populated from the application config, but they are not part of the application instance.

The application in the codeception module is recreated for every test, but the container is not. As @DmLapin correctly notes, the module uses the container to ask for an application object given a configuration array. If this array refers to objects that are not yet configured, because they are configured by the application at some later stage this will cause issues.

I see several options to remedy the situation:
1. Revert the change from #18127 
2. Refactor the config so DI container config lives outside the application config, which from my point of view will improve the code structure regradless of other solutions
3. Have the Yii2 Codeception module parse the configuration array and treat some keys specially  


I think 3 is the worst solution; the whole point of the module is to NOT know how the configuration format works, it is not relevant if we let Yii's own code figure that out. I believe 2 is a bad thing to ask from a user even if it (subjectively, in my view) improves their code; especially considering Yii2 should not be releasing BC breaking versions.

The option 1 is best in my opinion, since we have only had the feature for a very short while there will not be a lot of applications that are not under active development that use it. Apps under active development that use such a new feature will likely be able to step away from it. 
Upon further inspection, I think the `resolveDependencies` call should also be more secure in things like forwarding the `$reflection` variable, since it makes no sense to use that for nested array dependencies.. 

In conclusion, I propose reverting #18127 and for these advanced cases just relying on a `\Closure` for configuration of the DI container. It is cleaner and more readable for all PHP users instead of using some secret array language understood only by Yii developers. The array config should be limited to simple cases in all other cases it imo more sensible to use a (static) closure.

**Comment by samdark** [comment]
@hiqsol what do you think about reverting it?

**Comment by DmLapin** [comment]
Additional thoughts about Codeception connector. If recursive `resolveDependencies` will be remained (now or in future), it will be not good way to create application object through `Yii::createObject`. It will try to resolve all dependencies at start but not on-demand. Performance of tests will degrade.

**Comment by SamMousa** [comment]
It won't, the dependencies need to be instantiated anyway, so delaying it is not an option.
Note that optional dependencies still are delayed (not all components are instantiated for example). So performance wise it is not an issue.

**Comment by DmLapin** [comment]
> It won't, the dependencies need to be instantiated anyway, so delaying it is not an option.

If one test use only one application component, we shouldn't resolve all `Instance` objects on all levels of application config - it's waste of time. See example in my first post: `NewsBorrowCommandTest` don't use `internalFilesystem` and don't even use `NewsletterEmailGeneratorInterface`.

**Comment by hiqsol** [comment]
Sorry for long delay. I had no time.
I can't reproduce the issue. I'm trying to create a test for this issue.
Please see #18270 .
What needs to be added to reproduce the problem?

**Comment by samdark** [comment]
@DmLapin, @SamMousa would you please help with some additional info?

**Comment by samdark** [comment]
@SamMousa would it solve the issue to create application with `new` in Codeception module same way as it is done in framework bootstrap file itself?

**Comment by SamMousa** [comment]
Sure but that will break tests for other people. The issue with this is that it breaks BC and that simply means should revert it...

**Comment by SamMousa** [comment]
It would break for people who have subclassed the application object..

**Comment by samdark** [comment]
Added a test for it.
